lib/spack/spack/filesystem_view.py

```diff
@@ -40,6 +41,44 @@
 _projections_path = '.spack/projections.yaml'
 
 
+def view_symlink(src, dst, view, spec):
+    os.symlink(src, dst)
+
+
+def view_hardlink(src, dst, view, spec):
+    os.link(src, dst)
+
+
+def view_copy(src, dst, view, spec):
+    shutil.copyfile(src, dst)
+    if spec:
+        # Not metadata, we have to relocate it
+
+        # What type of file are we relocating
+        relocate_method = spack.relocate.relocate_text_bin \
+            if spack.relocate.is_binary(dst) else spack.relocate.relocate_text
+
+        # Get information on where to relocate from/to
+        prefix_to_prefix = dict(
+            (dep.prefix, view.get_projection_for_spec(dep))
+            for dep in spec.traverse()
+        )
+
+        # Call actual relocation method
+        relocate_method(
+            [dst], spack.store.layout.root, view._root,
+            spec.prefix, view.get_projection_for_spec(spec),
+            spack.paths.spack_root, view._root, prefix_to_prefix
+        )
+
+
+def view_link_wrapper(view, link_method):
+    def view_link(src, dst, spec=None):
+        link_method(src, dst, view, spec)
+
+    return view_link
```

Comment:
I think we can use `functools.partial` instead of this wrapper.

Reply:
This proved tricky, because of the way the wrapper was setting the default for the spec argument (which can't be set in the function signature, because `view` is a mandatory argument.

I couldn't swap the order, because then the keyword argument `view` comes before the argument `spec` once I call `functools.partial`.

The only thing that worked was to make both into keyword arguments. I'm perfectly okay with that, and I've done so.